Wrappers/Python/cil/optimisation/functions/IndicatorBox.py

```diff
@@ -29,67 +35,389 @@ class IndicatorBox(Function):
+                Output of the proximal operator. If not provided, a new DataContainer is created.
+
         '''
 
+        # x - tau * self.proximal(x/tau, tau)
+        should_return = False
+        
+        if out is None:
+            out = self.proximal(x, tau)
+            should_return = True
+        else:
+            self.proximal(x, tau, out=out)
+        
+        out.sapyb(-1., x, 1., out=out)
+
+        if should_return:
+            return out
+
+    def proximal(self, x, tau, out=None):
+        
+        r'''Proximal operator of IndicatorBox at x
+
+        .. math:: prox_{\tau * f}(x)
+
+        Parameters
+        ----------
+
+        x : DataContainer
+            Input to the proximal operator
+        tau : float
+            Step size. Notice it is ignored in IndicatorBox
+        out : DataContainer, optional
+            Output of the proximal operator. If not provided, a new DataContainer is created.
+
+        Note
+        ----
+
+            ``tau`` is ignored but it is in the signature of the generic Function class
+        '''
+        should_return = False
         if out is None:
+            should_return = True
+            out = x.copy()
+        else:
+            out.fill(x)
+        outarr = out.as_array()
+
+        # calculate the proximal
+        self._proximal(outarr)
+        
+        out.fill(outarr)
+        if should_return:
+            return out
+
+    def gradient(self,x):
+        '''IndicatorBox is not differentiable, so calling gradient will raise a ``ValueError``'''
+        return ValueError('Not Differentiable') 
+
+    def _proximal(self, outarr):
+        raise NotImplementedError('Implement this in the derived class')
+
+class IndicatorBox_numba(IndicatorBox):
+
+    def evaluate(self,x):
+        
+        '''Evaluates IndicatorBox at x'''
+        
+        num_threads = numba.get_num_threads()
+        numba.set_num_threads(cil_mp.NUM_THREADS)
+        breaking = np.zeros(numba.get_num_threads(), dtype=np.uint8)
+                
+        if isinstance(self.lower, np.ndarray):
+            if isinstance(self.upper, np.ndarray):
+
+                _array_within_limits_aa(x.as_array(), self.lower, self.upper, breaking)
+
+            else:
+
+                _array_within_limits_af(x.as_array(), self.lower, self.upper, breaking)
+
+        else:
+            if isinstance(self.upper, np.ndarray):
+
+                _array_within_limits_fa(x.as_array(), self.lower, self.upper, breaking)
+
+            else:
+
+                _array_within_limits_ff(x.as_array(), self.lower, self.upper, breaking)
+
+        numba.set_num_threads(num_threads)
+        return np.inf if breaking.sum() > 0 else 0.0
+    
+    def convex_conjugate(self,x):
+        '''Convex conjugate of IndicatorBox at x'''
+        # set the number of threads to the number of threads used in the CIL multiprocessing module
+        num_threads = numba.get_num_threads()
+        numba.set_num_threads(cil_mp.NUM_THREADS)
```

Comment:
could this be set globally?

Reply:
I was thinking that we may want to have an internal member (defaults to `NUM_THREADS`) we can use to set the number of threads in this class:

```
ib = IndicatorBox(lower, upper)
ib.set_num_threads(num_threads)
```